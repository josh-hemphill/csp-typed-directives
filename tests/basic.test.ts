import { Directives, ReportTo } from '../src/csp.types.js';
import { CspDirectives } from '../src/index.js';
import { createHash } from 'crypto';
const sample64Hash = (algorithm: string) => {
	const hash = createHash(algorithm);
	hash.update('hello world');
	return hash.digest('base64');
};

describe('new CspDirectives()',() => {
	it('Instantiates', () => {
		const inst = jest.fn(() => new CspDirectives());
		const res = inst();
		expect(inst).toHaveBeenCalled();
		expect(inst).toHaveReturned();
		expect(res).toMatchObject({
			CSP:{},
			ReportOnly: false,
			ReportTo: [],
			ReferrerHeader: 'strict-origin-when-cross-origin',
		});
	});
	it('Defaults Referer to deprecated referrer directive if present',() => {
		const inst = new CspDirectives({
			'referrer': 'strict-origin',
		});
		expect(inst.ReferrerHeader).toBe('strict-origin');
	});
	describe('.headers',() => {
		it('returns on empty',() => {
			const inst = new CspDirectives();
			const getHeaders = jest.spyOn(inst,'getHeaders');
			const headers = inst.getHeaders();
			expect(getHeaders).toHaveReturned();
			expect(headers).toMatchObject({
				'Content-Security-Policy-Report-Only': '',
				'Content-Security-Policy': '',
				'Report-To': '',
				'Referrer-Policy': 'strict-origin-when-cross-origin',
			});
		});
		it('returns boolean directives',() => {
			const csp: Directives = {
				'upgrade-insecure-requests': true,
			};
			const inst = new CspDirectives(csp,[],csp);
			const getHeaders = jest.spyOn(inst,'getHeaders');
			const headers = inst.getHeaders();
			expect(getHeaders).toHaveReturned();
			expect(headers).toMatchObject({
				'Content-Security-Policy-Report-Only': 'upgrade-insecure-requests;',
				'Content-Security-Policy': 'upgrade-insecure-requests;',
				'Report-To': '',
				'Referrer-Policy': 'strict-origin-when-cross-origin',
			});
		});
		it('returns on all set',() => {
			const sampleSha256 = `sha256-${sample64Hash('sha256')}` as const;
			const csp: Directives = {
				'child-src': 'none',
				'default-src': 'self',
				'frame-src': 'unsafe-eval',
				'connect-src': 'example.com',
				'font-src': 'https:',
				'img-src': 'self',
				'manifest-src': 'https://example.com',
				'media-src': sampleSha256,
				'object-src': 'example.com:443',
				'prefetch-src': sampleSha256,
				'script-src': 'strict-dynamic',
				'script-src-elem': sampleSha256,
				'script-src-attr': sampleSha256,
				'style-src': sampleSha256,
				'style-src-elem': sampleSha256,
				'style-src-attr': sampleSha256,
				'base-uri': 'strict-dynamic',
				'form-action': 'report-sample',
				'frame-ancestors': 'self',
				'navigate-to': sampleSha256,
				'report-to': 'hello',
				'referrer': 'strict-origin',
				'trusted-types': sampleSha256,
			};
			const endpoint = 'https://example.com' as const;
			const reportTo: ReportTo[] = [
				{
					max_age: 12000,
					group: 'hello',
					endpoints: [{url:endpoint}],
				},
			];
			const inst = new CspDirectives(csp,reportTo,csp,'strict-origin');
			const getHeaders = jest.spyOn(inst,'getHeaders');
			const headers = inst.getHeaders();
			expect(getHeaders).toHaveReturned();
			expect(headers).toMatchSnapshot();
		});
		it('Throws on invalid "report-to" group name',() => {
			const inst = new CspDirectives({
				'report-to': 'invalid',
			});
			jest.spyOn(inst,'getHeaders');
			try {
				inst.getHeaders();
			} catch (_) {
				'nothing';
			}
			expect(inst.getHeaders).toThrowError();
		});

		it('supports wildcards',() => {
			const csp: Directives = {
				'style-src': ['*', 'data:'],
			};
			const inst = new CspDirectives(csp, [], csp);
			expect(inst.getHeaders()).toStrictEqual({
				'Content-Security-Policy-Report-Only': 'style-src * data:;',
				'Content-Security-Policy': "style-src * data:;",
				'Report-To': '',
				'Referrer-Policy': 'strict-origin-when-cross-origin',
			});
		});
	});
});
