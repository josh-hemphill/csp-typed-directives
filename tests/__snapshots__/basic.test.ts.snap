// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new CspDirectives() .headers returns on all set 1`] = `
Object {
  "Content-Security-Policy": "child-src 'none'; default-src 'self'; frame-src 'unsafe-eval'; connect-src example.com; font-src https:; img-src 'self'; manifest-src https://example.com; media-src 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; object-src example.com:443; prefetch-src 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; script-src 'strict-dynamic'; script-src-elem 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; script-src-attr 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; style-src 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; style-src-elem 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; style-src-attr 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; base-uri 'strict-dynamic'; form-action 'report-sample'; frame-ancestors 'self'; navigate-to 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; report-to hello; referrer 'strict-origin'; trusted-types 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek=';",
  "Content-Security-Policy-Report-Only": "child-src 'none'; default-src 'self'; frame-src 'unsafe-eval'; connect-src example.com; font-src https:; img-src 'self'; manifest-src https://example.com; media-src 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; object-src example.com:443; prefetch-src 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; script-src 'strict-dynamic'; script-src-elem 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; script-src-attr 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; style-src 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; style-src-elem 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; style-src-attr 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; base-uri 'strict-dynamic'; form-action 'report-sample'; frame-ancestors 'self'; navigate-to 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek='; report-to hello; referrer 'strict-origin'; trusted-types 'sha256-uU0nuZNNPgilLlLX2n2r+sSE7+N6U4DukIj3rOLvzek=';",
  "Referrer-Policy": "strict-origin",
  "Report-To": "[{\\"max_age\\":12000,\\"group\\":\\"hello\\",\\"endpoints\\":[{\\"url\\":\\"https://example.com\\"}]}]",
}
`;
